### ISSUE TITLE
QTE's - hold instead of mash

### ISSUE THREAD

**Issue by nr1995** [nipkownix/re4_tweaks#107]
How about an option
to just hold the button instead of having to mash it during QTE's
for those of us with sore joints?
Thanks!

**Comment by Mister-Curious** [comment]
[AutoHotKey](https://www.autohotkey.com/) is your best bet here. You can use it to spam keyboard strokes, and it can be used in conjunction with [DS4Windows](https://ryochan7.github.io/ds4windows-site/) if you are using a controller. There are lots of threads in their forums about how to assign keystrokes to spam keys.

**Comment by nr1995** [comment]
It's not a baaaad idea... buuut...
Auto-fire through AHK via the controller doesn't work very well,
and auto-fire through AHK via the keyboard means I have to move hands from controller to keyboard,

On my GameCube-style controller, I'm only ever mashing the A or B button.
But it can't just be “hold A to mash A” or “hold B to mash B”
because then it'd always activate,
and I tried to make it only mash A or B while I'm holding down the L1/LB button,
but then the game would have to register L1/LB + a keyboard key,
and the game doesn't support simultaneous K+M/controller input.
It really is a pain.

**Comment by emoose** [comment]
It might be possible by patching `KeyTrgCheck` calls to use `KeyRepCheck` instead - AFAIK `Trg` (Triggered) returns true when button state was _changed_ to pressed from unpressed state, and returns false is button state hasn't changed (eg. if it's being held down) - while `Rep` (Repeat) returns true while button is being held down.

Looking at some of the QTEs that re4_tweaks has patches for already, it looks like some of them are using `KeyTrgCheck`, but some are also checking the triggered variable by themselves instead, so unfortunately patching `KeyTrgCheck` to always call `KeyRepCheck` probably wouldn't be enough (and would probably break some things too) - so I guess main way to allow this would be to go through code for each QTE & manually patch each one...

**Comment by nipkownix** [comment]
> Looking at some of the QTEs that re4_tweaks has patches for already, it looks like some of them are using KeyTrgCheck, but some are also checking the triggered variable by themselves instead

Something I've been thinking about doing is revisiting the QTE patches, now that we have your IDB file as a reference. Back when I first made those patches, I had no idea what I even was patching, tbh. 
The current approach probably wouldn't scale well with above 60 FPS too, so yet another reason to reimplement this stuff.
I'll probably take a look at it again soon.

**Comment by nipkownix** [comment]
Added an option to disable the QTEs, like modern games.

Here's a test build:
[dinput8-disableQTE_v1.zip](https://github.com/nipkownix/re4_tweaks/files/8164382/dinput8-disableQTE_v1.zip)

It makes the QTEs pass automatically and disables the on screen prompts, making QTEs look more like cutscenes.

_Apparently_ everything works fine, but I'd appreciate some testing.

**Comment by nr1995** [comment]
@nipkownix
This is only for occasions where you're asked to mash a button really fast right?
Not for any A+B or L+R or whatever, right?
If so, it's not a bad solution.
If not, it's definitely less than ideal.

**Comment by nipkownix** [comment]
Nope, this is for all QTEs, not just the mashing ones.

I didn't think people cared about the other ones, tbh. Wouldn't be hard to add another option to make just the mashing ones automatic, I guess.

**Comment by nr1995** [comment]
@nipkownix 
I'm just trying to detract as little as possible from the game's character.
If it's not hard to make a mashing-only one, I'd humbly ask for that.

**Comment by nipkownix** [comment]
New build:
[dinput8-disableQTE_v2.zip](https://github.com/nipkownix/re4_tweaks/files/8164904/dinput8-disableQTE_v2.zip)

> ; Disables most of the QTEs, making them pass automatically.
> DisableQTE = false
> 
> ; Unlike the previous option, this only automates the "mashing" QTEs, making them pass automatically. Prompts are still shown!
> AutomaticMashingQTE = false

@nr1995 let me know if that works for you.

**Comment by nipkownix** [comment]
You can enable the new settings using the F1 menu

**Comment by nr1995** [comment]
@nipkownix  Oh right.  Dumb.  Sorry.

**Comment by nr1995** [comment]
@nipkownix 
Yeah it's a bit janky in the laser corridor, I've just discovered.
Doesn't seem to work great in there.
